### DISCUSSION TITLE
Dependency Injection - Inject child class

### DISCUSSION THREAD

**Discussion by Darkheir** [fastapi/fastapi#8764] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆
### Description

I there a way to specify to inject a child class instead of the class declared as a dependency ?

In my example I want to specify that all my endpoints must be authenticated, but I don't want to have to specify how the authentication must be done at the endpoint declaration level.

I would like to be able to specify that when `Auth` is indicated as a dependency then `BasicAuth` must be injected.

This way changing the authentication medium would need to change a single configuration line and not all the endpoints.

### Example Code

```python
from starlette.requests import Request

class Auth:
    """
    Base class that all authenticator will extend
    """
    def __init__(self, request: Request):
        self.perform_authentication(request)

    def perform_authentication(self, request: Request):
        raise NotImplementedError()


class BasicAuth(Auth):
    def perform_authentication(self, request: Request):
        # Perform auth based on the HTTP Basic header


# I want my endpoint to be authenticated but I don't care about the way here
@app.get("/items/", dependencies=[Depends(Auth))
async def read_items():
    pass
```

### Operating System

Linux

### FastAPI Version

0.67

### Python Version

3.10

**Comment by Arpafaucon** [comment]
Hi! 
I am interested as well by this question. My use case is similar : were building an app that will ultimately be authenticated with OpenID, but we'd rather have a simpler scheme when testing (e.g. A basic Auth with hard-coded credentials), to avoid setting up a whole authentication test stack every time. 
I hope we can find a solution together :) 
I have done some small projects with FastAPI, and for now I believe something could be done using `app.depency_overrides`.I haven't set up a POC yet,though.

**Comment by chbndrhnns** [comment]
Would an approach like below work for your scenario?

```py
from functools import lru_cache

import pytest
from fastapi import FastAPI, Depends
from starlette.requests import Request
from starlette.testclient import TestClient


class Auth:
    """
    Base class that all authenticator will extend
    """

    def __call__(self, request: Request):
        raise Exception()


class BasicAuth(Auth):
    def __call__(self, request: Request):
        return True


my_app = FastAPI()
auth = Auth()


# I want my endpoint to be authenticated but I don't care about the way here
@my_app.get("/items/", dependencies=[Depends(auth)])
async def read_items():
    pass


class Tests:

    @pytest.fixture
    def app(self):
        return my_app

    @pytest.fixture
    def client(self, app):
        with TestClient(app=app) as c:
            yield c

    def test_no_override(self, client, app):
        with pytest.raises(Exception):
            client.get("/items/")

    def test_override(self, client, app):
        app.dependency_overrides[auth] = BasicAuth()

        res = client.get("/items/")
        assert res.status_code == 200
```

**Comment by Arpafaucon** [comment]
Hi, thanks for your code snippet! That's along the lines of what I had in mind. 

The downside is that, when doing it this way, you don't provide any information about your authentication scheme to fastapi, so `openapi.json` (and then swagger UI) won't be able to document it properly.

I must admit I really love the built-in authentication features inside Swagger, so I am desperately trying to find a solution that preserves it. Maybe that's irrealistic :/

**Comment by Arpafaucon** [comment]
I did some testing, and I kinda think that the root of the problem is that we want to mix `import`-time (when the `Depends` is created) and runtime behaviours (when the server starts).  

From what I understand, the analysis of dependencies is mostly done at import-time. When this snippet is imported:
```py
router = APIRouter()

@router.get("/help")
def help(a = Depends(auth))
```
the router stores a reference to the `auth` object that was provided, and it's by analyzing its arguments that the `FastAPI` dependency injection + openAPI schema build magic works. 

For me, this magic is so nice (kudos to the devs!!) that I am not ready to live without it.

So I think I will stick to a more `flask`-y factory pattern where routes are generated through factory functions. Snippet below:

```py
from fastapi import Depends
from fastapi.applications import FastAPI
from fastapi.routing import APIRouter
from fastapi.security import OAuth2PasswordBearer, HTTPBasic, HTTPBasicCredentials
from dataclasses import dataclass
import typing as tp


@dataclass
class UserData:
    name: str

AuthScheme = tp.Callable[..., UserData]

## define routes

def router(auth: AuthScheme) -> APIRouter:
    router = APIRouter()

    @router.get("/")
    def test(auth: UserData =Depends(auth)):
        return "hello " + auth.name

    return router

## configure multiple auth scheme. We'll chose among them at runtime

# debug scheme
httpbasic = HTTPBasic()
def auth1(basic_auth: HTTPBasicCredentials = Depends(httpbasic)):
    """Simple auth scheme for debugging"""

    return UserData(name=basic_auth.username)

# Production scheme
oauth = OAuth2PasswordBearer(tokenUrl="token")
def auth2(token: str = Depends(oauth)):
    # extract user info from token, build a UserData out of it
    return UserData(name="toto")


def build_app(debug_auth: bool):
    app = FastAPI()
    if debug_auth:
        auth = auth1
    else:
        auth = auth2

    app.include_router(router(auth))
    return app


if __name__ == "__main__":
    app = build_app(False)
    import uvicorn

    uvicorn.run(app, port=8000)
```

In this example, I have a weak inheritance pattern: the `router` factory expect a security dependency with a specific signature, and will use it to build the routes. Since we're not building anything at import time, we have time to create the authentication scheme we want (according to a config file, for example), and build the routes *afterwards* with it.

The code is slightly uglier since you have to wrap all your routes inside a function, though. For me, that's an acceptable price to pay (but I would delighted if someone finds anything better, of course :smile: )

**Comment by adriangb** [comment]
This is a normal use case for DI. What you want is to declare a dependency on an interface/abstraction and then bind a concrete implementation for deployment or testing. I think #3641 takes care of this use case since you can `.bind(Auth, BasicAuth)`

**Comment by Darkheir** [comment]
> This is a normal use case for DI. What you want is to declare a dependency on an interface/abstraction and then bind a concrete implementation for deployment or testing. I think #3641 takes care of this use case since you can `.bind(Auth, BasicAuth)`

Yes this is typically what I would like !
From what I understood it is not possible as of now, but I will watch closely your PR :-)
